In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
path=r'C:\Users\olilo\OneDrive\Dokumente\CF\Data Analytics Immersion_Week 8-9\2023-08-28 Instacart Basket Analysis'

In [4]:
#Create a dataframe

df_test = pd.DataFrame()

In [5]:
#Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [6]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [7]:
#Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [8]:
df_test['mix'] = df_test['mix'].astype('str')

In [9]:
#“Original Data-“products.csv” dataframe

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [10]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
df_nan=df_prods[df_prods['product_name'].isnull()== True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_prods.shape

(49693, 5)

In [14]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
df_prods_clean.shape

(49677, 5)

In [16]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

In [21]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [22]:
# Add orders.csv from Original Data

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Count of rows in all columns is the same. 
There is a discrepancy between the minimum and maximum by column days_since_prior_order. 
Min = 0 days
Max = 3 days
but 50% they are valid for 7 days. 
After a second look, however, I find that everything is correct when the ending sproves that the max is not 3 but 30 (3.000000e+01). So, 7.0 * 10^0 is equal to 7.0, and 3.0 * 10^1 is equal to 30.0.
Max = 30 days.

In [23]:
#Check for mixed types of data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

The code gives no output, so there are no mixed types of data

In [24]:
#Check for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In this column, days_since_prior_order the code count 206209 missing values or 0. It is logical that there are zero days since the order was placed, because at this moment it can be hours or minutes.
The column has criteria of days, so I'm not going to delete them, as my changes will appear every second, when more than 24 hours have passed since the time of ordering.
At the same time, partly by cilumne orders_dow , we can conclude that some orders are over 24 hours old.
To consider this task in its own way, you would need to write two python codes. One finding missing data in column days_since_prior_order at the same time as the number from 0 to 7 corresponds to today's day in the order_dow column. And the second code taking all other day from column order_dow and NaN in the column days_since_prior_order.

The Python approximation is not sufficient to achieve this at the moment, so create a new dataframe only with rows with NaN. 
And decides to leave the missing data for the moment. In real life, I would go with a question or request for clarification to my client.

In [25]:
df_onan=df_ords[df_ords['days_since_prior_order'].isnull()== True]

In [26]:
df_onan.shape

(206209, 7)

In [27]:
df_onan.head(30)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0     2539329        1    prior             1          2                  8   
11    2168274        2    prior             1          2                 11   
26    1374495        3    prior             1          1                 14   
39    3343014        4    prior             1          6                 11   
45    2717275        5    prior             1          3                 12   
50    2086598        6    prior             1          5                 18   
54    2565571        7    prior             1          3                  9   
75     600894        8    prior             1          6                  0   
79     280530        9    prior             1          1                 17   
83    1224907       10    prior             1          2                 14   
89    2411567       11    prior             1          0                 11   
97    1483446       12    prior             1          1                 12   
103   2618231       13    prior             1          6                 12   
116   2906263       14    prior             1          2                 18   
130   2514324       15    prior             1          5                 13   
153   1273676       16    prior             1          2                  8   
160   1737705       17    prior             1          2                 13   
201   2780889       18    prior             1          6                 18   
208   1321974       19    prior             1          6                 13   
218    947722       20    prior             1          1                 10   
223   2625565       21    prior             1          1                 10   
257   2300537       22    prior             1          3                  7   
273   1792521       23    prior             1          5                 14   
278    611214       24    prior             1          3                 20   
297   2610576       25    prior             1          1                 14   
301   3229714       26    prior             1          2                 15   
314    361493       27    prior             1          3                  9   
396   1138968       28    prior             1          0                 14   
421   2064090       29    prior             1          5                 11   
440    375303       30    prior             1          1                 13   

     days_since_prior_order  
0                       NaN  
11                      NaN  
26                      NaN  
39                      NaN  
45                      NaN  
50                      NaN  
54                      NaN  
75                      NaN  
79                      NaN  
83                      NaN  
89                      NaN  
97                      NaN  
103                     NaN  
116                     NaN  
130                     NaN  
153                     NaN  
160                     NaN  
201                     NaN  
208                     NaN  
218                     NaN  
223                     NaN  
257                     NaN  
273                     NaN  
278                     NaN  
297                     NaN  
301                     NaN  
314                     NaN  
396                     NaN  
421                     NaN  
440                     NaN

In [28]:
#Looking for full duplicate values
df_ords.shape

(3421083, 7)

In [29]:
df_odups = df_ords[df_ords.duplicated()]

In [30]:
df_odups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [31]:
df_odups.shape

(0, 7)

No duplicates found in this dataframe

In [32]:
#Save the new dataset
df_odups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [33]:
#Save the new dataset without duplication
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked_nodupli.csv'))